### ISSUE TITLE
Can't build my package.yml (3 different errors)

### ISSUE THREAD

**Issue by magnusviri** [teaxyz/pantry.core#17]
I am trying to build a project in a Docker container and I get this error.

```
TEA_PREFIX=/opt /opt/tea.xyz/var/pantry/scripts/build.ts isc.org/dhcp
error: Uncaught Error: not-found:srcroot
        throw new Error("not-found:srcroot");
              ^
    at file:///opt/tea.xyz/src/v0.11.4/src/hooks/useVirtualEnv.ts:34:15
    at useVirtualEnv (file:///opt/tea.xyz/src/v0.11.4/src/hooks/useVirtualEnv.ts:35:7)
    at abracadabra (file:///opt/tea.xyz/src/v0.11.4/src/app.exec.ts:142:25)
    at async exec (file:///opt/tea.xyz/src/v0.11.4/src/app.exec.ts:12:57)
    at async file:///opt/tea.xyz/src/v0.11.4/src/app.ts:29:9
```

Steps to recreate:

    git clone https://github.com/magnusviri/pantry.magnusviri
    cd pantry.magnusviri
    docker build -t tea .
    docker run --rm -it -v $PWD/projects/isc.org:/opt/tea.xyz/var/pantry/projects/isc.org tea

In the container this produces the above error.

    TEA_PREFIX=/opt /opt/tea.xyz/var/pantry/scripts/build.ts isc.org/dhcp
   
This works more.

    TEA_DIR=/opt/tea.xyz/var/pantry TEA_PREFIX=/opt /opt/tea.xyz/var/pantry/scripts/build.ts isc.org/dhcp

It produces this error.

```
tea: installed: /opt/gnu.org/tar/v1.34.0
tea: installed: /opt/tukaani.org/xz/v5.2.7
tea: installed: /opt/sourceware.org/bzip2/v1.0.8
tea: installed: /opt/deno.land/v1.27.0
error: Unable to load 'file:///opt/tea.xyz/var/pantry/import-map.json' import map

Caused by:
    No such file or directory (os error 2)
```

Debugging:

```
# ls -l /opt/tea.xyz/var/pantry/import-map.json
# lrwxrwxrwx 1 root root 22 Nov  3 18:11 /opt/tea.xyz/var/pantry/import-map.json -> ../cli/import-map.json
# find /opt/ -name cli
# find /opt/ -name import-map.json
/opt/tea.xyz/var/pantry/import-map.json
# ls -l /opt/tea.xyz/var/pantry/import-map.json
lrwxrwxrwx 1 root root 22 Nov  3 18:11 /opt/tea.xyz/var/pantry/import-map.json -> ../cli/import-map.json
```

If I put the cli in the project directory and launch the container like this:

    docker run --rm -it -v $PWD/projects/isc.org:/opt/tea.xyz/var/pantry/projects/isc.org -v $PWD/cli:/opt/tea.xyz/var/cli tea

then 

    TEA_DIR=/opt/tea.xyz/var/pantry TEA_PREFIX=/opt /opt/tea.xyz/var/pantry/scripts/build.ts isc.org/dhcp

produces this error:

```
error: Uncaught (in promise) Error
  throw new Error(message)
        ^
    at panic (file:///opt/tea.xyz/var/pantry/src/utils/index.ts:246:9)
    at file:///opt/tea.xyz/var/pantry/src/hooks/usePantry.ts:214:42
    at Array.map (<anonymous>)
    at getVersions (file:///opt/tea.xyz/var/pantry/src/hooks/usePantry.ts:213:21)
    at async Object.resolve (file:///opt/tea.xyz/var/pantry/src/hooks/usePantry.ts:36:20)
    at async file:///opt/tea.xyz/var/pantry/scripts/build.ts:45:15
```

At this point it's a little over my head.

**Comment by mxcl** [comment]
Sorry about this @magnusviri, certainly the steps to build are more finicky that they should be. I'm actively trying to improve this currently.

Good sleuthing! You really figured out what tea/cli was looking for. I'm actively working on improving the README and fixing some of these issues RN.

The code is throwing at

```
semver.parse(validate_str(raw)) ?? panic()
```

So basically it's a bad error message, going to commit a fix for that immediately.

It failed to parse the versions from isc.org. Care to post the `package.yml`?

**Comment by mxcl** [comment]
Oh you also need to copy the scripts from here. lol

Let me spend some more time figuring out how to have this work

**Comment by mxcl** [comment]
k merged so you don't need to build in `/opt` anymore (but we do in CI for consistency)

The rules currently still are:

1. clone `tea/cli`, `panty.core` and your pantry alongside each other
2. run `TEA_PANTRY_PATH=/ur/pantry pantry.core/scripts/build.ts`

That should now work. Got plans for reducing this to a one liner but can't do that in a couple hours, needs a bit more work.

**Comment by magnusviri** [comment]
That worked a little better but it doesn't seem to have run any of the script. Tea's dir is /opt, and my pantry is /opt/build/pantry. [Here is the package.yml](https://github.com/magnusviri/pantry.magnusviri/blob/main/projects/isc.org/dhcp/package.yml).

```
root@1f765748eedc:/# ls -l /opt/build/
total 0
drwxr-xr-x 14 root root 448 Nov  5 06:12 cli
drwxr-xr-x  9 root root 288 Nov  5 06:12 pantry
drwxr-xr-x 11 root root 352 Nov  5 06:12 pantry.core
root@1f765748eedc:/# TEA_PANTRY_PATH=/opt/build/pantry /opt/build/pantry.core/scripts/build.ts
tea: installed: /opt/gnu.org/tar/v1.34.0
tea: installed: /opt/tukaani.org/xz/v5.2.7
tea: installed: /opt/sourceware.org/bzip2/v1.0.8
tea: installed: /opt/deno.land/v1.27.1
::set-output name=pkgs::
::set-output name=paths::
::set-output name=relative-paths::
::set-output name=srcs::
::set-output name=srcs-relative-paths::
root@1f765748eedc:/# find . -name dhcp
./opt/build/pantry/projects/isc.org/dhcp
root@1f765748eedc:/# ls -l /opt/build/pantry/projects/isc.org/dhcp
total 4
-rw-r--r-- 1 root root 517 Nov  5 06:24 package.yml
root@1f765748eedc:/#
```

**Comment by magnusviri** [comment]
Oh, I see, I left off the project to build. After adding it I get the same semver error. And it's a little better of an error message, although I wouldn't have guessed that "semver" meant semantic version. But since semver.org exists, I'm guessing I'm in the minority of devs that wouldn't know that.

I went and moved the part of the version that didn't match the regex from the versions array to this.

```
distributable:
  url: https://downloads.isc.org/isc/dhcp/{{ version }}-P1/dhcp-{{ version }}-P1.tar.gz
```

Then when I tried to build it I got missing package errors like this.

```
error: Uncaught (in promise) Error: not-found:zlib.net^1
```

Here's the full list that I installed manually. I don't understand why these weren't found because the package.yml file specified most of these as build dependencies. But none of them installed.

- zlib.net
- llvm.org
- gnu.org/m4
- perl.org
- tea.xyz/gx/cc
- gnu.org/make
- gnu.org/autoconf
- gnu.org/automake
- gnu.org/libtool

After I installed all of them then it started to build and gave me an error that is probably because I'm doing something wrong in my package.yml file. I'll probably have to try to build it normally (without tea) to debug it, unless there's tea can add all of the deps to the enviroment. I don't know how to get tea to do that.

**Comment by mxcl** [comment]
k I've made builds less tedious, (still more to do, I want them to be as easy as poss ofc), but now you should just be able to run `build.ts foo.com` from pantry.core. See instructions here: https://github.com/teaxyz/pantry.zero#contributing

> although I wouldn't have guessed that "semver" meant semantic version

that's fair, I'll update it.

> Then when I tried to build it I got missing package errors like this.

The new build.ts will install the deps for you.

> I don't understand why these weren't found because the package.yml file specified most of these as build dependencies. But none of them installed.

The previous `build.ts` was “plumbing” rather “porcelain” and thus did not install deps for you. But the new one does.

Sorry for the hassle. Early days niggles!

> After I installed all of them then it started to build and gave me an error that is probably because I'm doing something wrong in my package.yml file. 

PR it and we'll help in there.

> I'll probably have to try to build it normally (without tea) to debug it, unless tea can easily add all of the deps to the environment. I don't know how to get tea to do that.

the build dir will be at `$TEA_PREFIX`/project/src/vx.y.z` with the build script in the parent directory to help with these issues.

**Comment by magnusviri** [comment]
I gave this another try. I understand it a little better and I was able to debug it to the point that I think the problem is that the app I'm trying to compile, isc.org/dhcp, doesn't build out-of-the-box. So I'm not sure if this is something that you guys can help with or not. I'm guessing it's not something you guys want to get involved with. So I'm probably going to try to find another package to attempt. I don't really need this anyway. I was just trying it. I will find something else to learn with.
